web/src/hotkey.js

```diff
@@ -163,107 +196,49 @@ const keydown_either_mappings = {
-    s: {name: "toggle_conversation_view", message_view_only: true},
-    t: {name: "open_recent_view", message_view_only: true},
-    u: {name: "toggle_sender_info", message_view_only: true},
-    v: {name: "show_lightbox", message_view_only: true},
-    w: {name: "query_users", message_view_only: true},
-    x: {name: "compose_private_message", message_view_only: true},
-    z: {name: "zoom_to_message_near", message_view_only: true},
+    "Shift+Backspace": {name: "backspace", message_view_only: true},
+    "Shift+Enter": {name: "enter", message_view_only: false},
+    "Shift+Delete": {name: "delete", message_view_only: false},
 };
 
 // Keyboard Event Viewer tool:
 // https://w3c.github.io/uievents/tools/key-event-viewer.html
 export function get_keydown_hotkey(e) {
-    let hotkey;
-
-    if (e.altKey) {
-        hotkey = keydown_alt_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
+    if (common.has_mac_keyboard() && e.ctrlKey && e.key !== "[") {
+        // On macOS, Cmd is used instead of Ctrl. Except 'Ctrl + ['.
         return undefined;
     }
-
-    if (e.ctrlKey && !e.shiftKey) {
-        hotkey = keydown_ctrl_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
-
-    const isCmdOrCtrl = common.has_mac_keyboard() ? e.metaKey : e.ctrlKey;
-    if (isCmdOrCtrl && !e.shiftKey) {
-        hotkey = keydown_cmd_or_ctrl_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-        return undefined;
-    } else if (e.metaKey || e.ctrlKey) {
+    if (!common.has_mac_keyboard() && e.metaKey) {
         return undefined;
     }
 
-    if (e.shiftKey) {
-        hotkey = keydown_shift_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
+    // Make lowercase a-z uppercase; leave others unchanged.
+    // We distinguish between 'a' and 'A' as 'A' vs 'Shift+A'.
+    // This helps to avoid Caps Lock affecting shortcuts.
+    let key = /^[a-z]$/.test(e.key) ? e.key.toUpperCase() : e.key;
 
-    if (!e.shiftKey) {
-        hotkey = keydown_unshift_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
+    // We ignore the modifiers for these keys as they are treated as
+    // involved in generating the character in various keyboard layouts.
+    // Here, modifiers are not the part of the shortcut key combination.
+    const ignore_modifiers_for_keys = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
 
-    hotkey = keydown_either_mappings[e.key];
-    if (hotkey) {
-        return hotkey;
+    if (!ignore_modifiers_for_keys.has(key)) {
+        if (e.shiftKey) {
+            key = "Shift+" + key;
+        }
+        if (e.altKey) {
+            key = "Alt+" + key;
+        }
+        if (e.ctrlKey) {
+            key = "Ctrl+" + key;
+        }
+        if (e.metaKey) {
+            assert(common.has_mac_keyboard());
+            key = "Cmd+" + key;
```

Comment:
Why is this assertion true? Does the spec say only mac keyboard systems can have a `metaKey`? I suspect this logic is wrong.

Reply:
Maybe we want this to be:
```
if (e.metaKey) {
    if (common.has_mac_keyboard()) {
        key = "MacCmd" + key;
    } else {
        key = "Cmd" + key;
    }
} 
```

and we'd have our CmdOrCtrl legacy keys just use `MacCmd` as the name -- so if you use a `Cmd` type key (I think the Windows key on non-Mac computers?) that doesn't fall into the Mac logic -- on a PC, you want `Ctrl` to do that and `Cmd` to not behave like `Ctrl` unexpectedly.

Reply:
> Why is this assertion true? Does the spec say only mac keyboard systems can have a metaKey? I suspect this logic is wrong.

For non-mac systems, our docs suggest to use `Ctrl` instead of `Cmd`. So, if `Cmd` is involved in a shortcut combination on a non-mac system, shouldn't we just return undefined ?

Reply:
```
    const isCmdOrCtrl = common.has_mac_keyboard() ? e.metaKey : e.ctrlKey;
    if (isCmdOrCtrl && !e.shiftKey) {
        hotkey = keydown_cmd_or_ctrl_mappings[e.key];
        if (hotkey) {
            return hotkey;
        }
        return undefined;
    } else if (e.metaKey || e.ctrlKey) {
        return undefined;
    }
    ```
    Earlier, this was our code. In case of mac, use `Cmd` otherwise `Ctrl`. If we're using `Meta` key on windows, that'd return `undefined`.

Reply:
This also resolves a bug as mentioned in commit message:
> This descriptive string approach also resolves a bug where
`Ctrl + Meta + C/K/S/.` was working. We define the valid hotkey as
`Meta + C/K/S/.` for macOS and `Ctrl + C/K/S/.` for others.